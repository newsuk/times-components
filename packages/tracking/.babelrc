{
  "presets": ["flow", "env", "react", "stage-2"]
}
