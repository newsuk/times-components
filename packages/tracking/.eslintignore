**/node_modules/*
**/coverage/*
dist
flow-typed
