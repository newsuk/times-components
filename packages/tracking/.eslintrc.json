{
  "extends": ["@times-components/thetimes"],
  "parser": "@babel/eslint-parser"
}
