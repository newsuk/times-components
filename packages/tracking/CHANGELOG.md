# Change Log

All notable changes to this project will be documented in this file.
See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.0.2"></a>
## 0.0.2 (2017-10-06)
