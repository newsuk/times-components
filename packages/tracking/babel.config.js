module.exports = api => {
  api.cache(true);
  return {
    plugins: [
      "babel-plugin-styled-components",
      "@babel/plugin-proposal-class-properties"
    ],
    presets: [
      [
        "@babel/preset-env",
        {
          targets: {
            node: "current"
          }
        }
      ],
      "@babel/preset-react",
      "@babel/preset-typescript"
    ]
  };
};
