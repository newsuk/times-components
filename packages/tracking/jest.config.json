{
  "preset": "react-native",
  "rootDir": "../../",
  "transformIgnorePatterns": ["node_modules/(?!@times-components)/"],
  "testMatch": ["<rootDir>/packages/tracking/__tests__/**.test.js"],
  "testPathIgnorePatterns": ["web.test.js$"]
}
