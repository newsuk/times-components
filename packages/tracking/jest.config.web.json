{
  "rootDir": "../../",
  "transformIgnorePatterns": ["node_modules/(?!@times-components)/"],
  "testEnvironment": "jsdom",
  "testMatch": ["<rootDir>/packages/tracking/__tests__/**.web.test.js"],
  "moduleNameMapper": {
    "react-native": "react-native-web"
  },
  "moduleFileExtensions": ["web.js", "js", "json"],
  "setupFiles": ["raf/polyfill"]
}
