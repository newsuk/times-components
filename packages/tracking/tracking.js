import React, { Component } from "react";
import hoistNonReactStatic from "hoist-non-react-statics";
import PropTypes from "prop-types";
import resolveAttrs from "./resolve-attrs";
import getDisplayName from "./get-display-name";

export { default as createConsoleReporter } from "./reporters/console-logger";
export { default as tealiumTransformer } from "./transformers/tealium";
export { default as addTrackingContext } from "./tracking-context";

export const trackingContextTypes = {
  tracking: PropTypes.shape({
    analytics: PropTypes.func
  })
};

export const addTracking = (
  WrappedComponent,
  { analyticsEvents = [], trackingName, attrs = {}, trackView } = {}
) => {
  const componentName = getDisplayName(WrappedComponent, trackingName);

  class WithTracking extends Component {
    constructor(props) {
      super(props);

      this.wrappedFuncs = new Map();
    }

    componentDidMount() {
      if (!this.context || !this.context.tracking) {
        return;
      }

      if (trackView) {
        this.context.tracking.analytics({
          object: componentName,
          action: "Rendered",
          attrs: {
            ...resolveAttrs(attrs, this.props)
          }
        });
      }
    }
    getWrappedAnalyticsEvents() {
      return this.wrapWithTracking(
        analyticsEvents,
        (resolvedAttrs, actionName) =>
          this.context.tracking &&
          this.context.tracking.analytics({
            object: componentName,
            action: "Action",
            attrs: {
              actionName,
              ...resolvedAttrs
            }
          })
      );
    }

    wrapWithTracking(eventNames, tracking) {
      return eventNames.reduce((wrappedFuncProps, funcName) => {
        const [origFunc, wrappedFunc] = this.wrappedFuncs.get(funcName) || [];

        if (wrappedFunc && this.props[funcName] === origFunc) {
          return {
            ...wrappedFuncProps,
            [funcName]: wrappedFunc
          };
        }

        const funcWrapped = (...args) => {
          tracking(resolveAttrs(attrs, this.props), funcName);
          return this.props[funcName] && this.props[funcName](...args);
        };

        this.wrappedFuncs.set(funcName, [this.props[funcName], funcWrapped]);

        return {
          ...wrappedFuncProps,
          [funcName]: funcWrapped
        };
      }, {});
    }

    render() {
      const wrappedProps = {
        ...this.props,
        ...this.getWrappedAnalyticsEvents()
      };

      return <WrappedComponent {...wrappedProps} />;
    }
  }

  WithTracking.displayName = `WithTracking(${componentName})`;

  hoistNonReactStatic(WithTracking, WrappedComponent);

  WithTracking.contextTypes = trackingContextTypes;
  WithTracking.propTypes = WrappedComponent.propTypes;
  WithTracking.defaultProps = WrappedComponent.defaultProps;

  return WithTracking;
};
