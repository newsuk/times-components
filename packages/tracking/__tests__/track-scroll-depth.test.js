import React from "react";
import renderer from "react-test-renderer";
import Enzyme, { shallow } from "enzyme";
import React16Adapter from "enzyme-adapter-react-16";
import { withTrackScrollDepth } from "../tracking";
import withTestContext from "./test-tracking-context";
import sharedTrackingTests from "./shared-tracking-tests";
import ListComponent from "./list-component";

Enzyme.configure({ adapter: new React16Adapter() });

describe("WithTrackScrollDepth", () => {
  it("tracks child views", () => {
    const reporter = jest.fn();
    const ListWithChildTracking = withTestContext(
      withTrackScrollDepth(ListComponent, {
        childIdPropKey: "someKey",
        getChildren: props => props.items
      }),
      { trackingObject: "TestObject" }
    );

    renderer.create(
      <ListWithChildTracking
        analyticsStream={reporter}
        onViewed={() => {}}
        items={[
          { someKey: "1", someValue: "one" },
          { someKey: "2", someValue: "two" },
          { someKey: "3", someValue: "three" }
        ]}
      />
    );

    expect(reporter.mock.calls).toMatchSnapshot();
  });

  it("accepts component name override", () => {
    const reporter = jest.fn();
    const ListWithChildTracking = withTestContext(
      withTrackScrollDepth(ListComponent, {
        childIdPropKey: "someKey",
        getChildren: props => props.items,
        trackingName: "SomeItem"
      })
    );

    const tracking = shallow(
      <ListWithChildTracking
        onViewed={() => {}}
        analyticsStream={reporter}
        items={[
          { someKey: "1", someValue: "one" },
          { someKey: "2", someValue: "two" },
          { someKey: "3", someValue: "three" }
        ]}
      />,
      {
        context: {
          tracking: {
            analytics: reporter
          }
        }
      }
    );
    tracking
      .dive()
      .dive()
      .instance()
      .onViewableItemsChanged.call(tracking.instance(), {
        info: {
          changed: [
            {
              isViewable: true,
              id: "one",
              someKey: "1",
              someValue: "one"
            }
          ]
        }
      });

    expect(reporter).toHaveBeenCalledWith(
      expect.objectContaining({
        component: "SomeItemChild"
      })
    );
  });

  it("accepts action name override", () => {
    const reporter = jest.fn();
    const ListWithChildTracking = withTestContext(
      withTrackScrollDepth(ListComponent, {
        childIdPropKey: "someKey",
        getChildren: props => props.items,
        actionName: "Scrolled"
      })
    );

    const tracking = shallow(
      <ListWithChildTracking
        onViewed={() => {}}
        analyticsStream={reporter}
        items={[
          { someKey: "1", someValue: "one" },
          { someKey: "2", someValue: "two" },
          { someKey: "3", someValue: "three" }
        ]}
      />,
      {
        context: {
          tracking: {
            analytics: reporter
          }
        }
      }
    );
    tracking
      .dive()
      .dive()
      .instance()
      .onViewableItemsChanged.call(tracking.instance(), {
        info: {
          changed: [
            {
              isViewable: true,
              id: "one",
              someKey: "1",
              someValue: "one"
            }
          ]
        }
      });
    expect(reporter).toHaveBeenCalledWith(
      expect.objectContaining({
        action: "Scrolled"
      })
    );
  });

  it("applies tracking attrs", () => {
    const reporter = jest.fn();
    const ListWithChildTracking = withTestContext(
      withTrackScrollDepth(ListComponent, {
        childIdPropKey: "someKey",
        getChildren: props => props.items,
        getAttrs: props => ({
          id: props.someKey,
          index: props.index,
          total: props.total
        })
      })
    );

    const tracking = shallow(
      <ListWithChildTracking
        onViewed={() => {}}
        analyticsStream={reporter}
        items={[
          { someKey: "1", someValue: "one" },
          { someKey: "2", someValue: "two" },
          { someKey: "3", someValue: "three" }
        ]}
      />,
      {
        context: {
          tracking: {
            analytics: reporter
          }
        }
      }
    );

    tracking
      .dive()
      .dive()
      .instance()
      .onViewableItemsChanged.call(tracking.instance(), {
        info: {
          changed: [
            {
              isViewable: true,
              id: "one",
              someKey: "1",
              someValue: "one"
            }
          ]
        }
      });
    expect(reporter).toHaveBeenCalledWith(
      expect.objectContaining({
        attrs: expect.objectContaining({ id: "1", index: 0, total: 3 })
      })
    );
  });

  it("can track scroll events", () => {
    const reporter = jest.fn();
    const ListWithChildTracking = withTestContext(
      withTrackScrollDepth(ListComponent, {
        childIdPropKey: "someKey",
        actionName: "Scrolled",
        getChildren: props => props.items,
        getAttrs: props => ({ depth: (props.index + 1) / props.total * 100 })
      })
    );

    const tracking = shallow(
      <ListWithChildTracking
        onViewed={() => {}}
        analyticsStream={reporter}
        items={[
          { someKey: "1", someValue: "one" },
          { someKey: "2", someValue: "two" },
          { someKey: "3", someValue: "three" }
        ]}
      />,
      {
        context: {
          tracking: {
            analytics: reporter
          }
        }
      }
    );
    tracking
      .dive()
      .dive()
      .instance()
      .onViewableItemsChanged.call(tracking.instance(), {
        info: {
          changed: [
            {
              isViewable: true,
              id: "three",
              someKey: "3",
              someValue: "three"
            }
          ]
        }
      });
    expect(reporter).toHaveBeenLastCalledWith(
      expect.objectContaining({
        action: "Scrolled",
        attrs: expect.objectContaining({ depth: 100 })
      })
    );
  });

  sharedTrackingTests(withTrackScrollDepth);
});
