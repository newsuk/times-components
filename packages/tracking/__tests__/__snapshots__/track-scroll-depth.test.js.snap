// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithTrackScrollDepth forwards props to wrapped component 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  bar
</Text>
`;

exports[`WithTrackScrollDepth renders when tracking context is missing 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  foo
</Text>
`;

exports[`WithTrackScrollDepth tracks scroll depth 1`] = `Array []`;
